reactor: support more network ops in io_uring backend

tchaikov opened the pull request:
* reorder `required_ops` in `try_create_uring` in the exact order of
  `io_uring_op` in linux/include/uapi/linux/io_uring.h. the
  `io_uring_op` enum is defined in the chronological order when the
  ops are added. and add comments to note the exact linux version
  in which the corresponding ops where added.
* move speculation from `reactor_backend_uring::poll()`, so we can
  be more explicit when calling `readable()` or `writeable()`. So,
  if `take_speculation()` is called, we want to have speculation
  explicitly. If `poll()` is called, the outcome should be a
  readable or writeable fd.
* document the `take_speculation()` and `speculate_epoll()` methods
* rewrite `reactor_backend_uring::read_some()` and
  `reactor_backend_uring::write_some()` with io_uring ops. please
  note, quote from the commit message of linux source commit d7718a9d

  > io_uring tries any request in a non-blocking manner, if it can,
  > and then retries from a worker thread if we get -EAGAIN.

  so, io_uring never returns EAGAIN, and after linux v5.7, it
  even polls for the events before performing I/O if -EAGAIN is
  returned. so we don't have to poll for the event or open the
  file / socket without `SOCK_NONBLOCK` or `O_NONBLOCK` in order
  to prevent io_uring from returning -EAGAIN.

tested on a 8-core x86 machine with `httpd --smp 1` (release build)
and `wrk -c 128 -t 4`. the aio backend showed:
```
Running 10s test @ http://localhost:10000/
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.49ms  109.56us   3.81ms   87.41%
    Req/Sec    21.51k   783.10    28.03k    87.84%
  862627 requests in 10.10s, 112.71MB read
Requests/sec:  85407.13
Transfer/sec:     11.16MB
```
while the io_uring backend gave us:
```
Running 10s test @ http://localhost:10000/
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.35ms  400.14us  37.64ms   98.36%
    Req/Sec    23.94k     2.36k   26.30k    70.25%
  952788 requests in 10.00s, 124.48MB read
Requests/sec:  95266.43
Transfer/sec:     12.45MB
```

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

tchaikov commented:
tested using httpd and wrk + liburing master HEAD + linux 6.0

tchaikov commented on src/core/reactor_backend.cc:
```diff
@@ -1463,29 +1489,142 @@ class reactor_backend_uring final : public reactor_backend {
+            return queue_request(internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false));
+        }).then([&fd, len] (size_t result) {
+            if (result == len) {
+                fd.speculate_epoll(EPOLLIN);
```
still need to find a way to handle `-EAGAIN` properly if we happen to run out of the read buffer and the last speculation fails. currently, `io_completion::complete_with()` translates `-EAGAIN` in `sqe->res` to a `std::system_error(EAGAIN, std::system_category())`.

avikivity replied:
We just have to loop through readable() again.

However, if reading without SOCK_NONBLOCK, it will never return EAGAIN.

tchaikov commented on src/core/reactor_backend.cc:
```diff
@@ -1342,9 +1353,17 @@ class reactor_backend_uring final : public reactor_backend {
+                break;
             case o::sendmsg:
+                ::io_uring_prep_sendmsg(sqe, req.fd(), req.msghdr(), req.flags());
+                break;
```
yeah, they have been around since uring 0.7.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1475,46 +1511,39 @@ class reactor_backend_uring final : public reactor_backend {
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, void* buffer, size_t len) override {
         return readable(fd).then([this, &fd, buffer, len] () mutable {
```
I think this pessimizes the single-connection case. readable() will return true (speculated), then we queue a request into the ring, but we can have latency of 0.5ms even though data is ready because the reactor will prefer to run tasks and gather more I/O. It's also suboptimal for many connections, since if we speculate that fd is not readable, we'll go through two trips into uring, one for readable() and one for queue_request().

I think we need this logic:

1. check speculation on fd
2. if we speculated that fd is readable, call fd.fd.read() to copy data immediately and return it
3. if we speculated that fd is not readable, call queue_request() immediately

For this to work well, we have to disable SOCK_NONBLOCK when using io_uring, and enable MSG_NOWAIT in step 2. With these changes, we'll have the following:

1. request/reply sockets (data usually not available) will usually wait in IORING_OP_RECV
2. streaming sockets will never wait and use recvmsg(..., MSG_NOWAIT) to copy data from the kernel to the buffer

tchaikov commented:
making this PR a draft as the speculation machinery does not work as expected.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1475,46 +1511,39 @@ class reactor_backend_uring final : public reactor_backend {
-            if (size_t(*r) == len) {
+            return queue_request(internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false));
+        }).then([&fd, len] (size_t result) {
+            if (result == len) {
```
You can merge the code here into the io_completion subclass. It's already a continuation, no need to add one more here. You'll probably need four subclasses, one for each case.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1475,46 +1511,39 @@ class reactor_backend_uring final : public reactor_backend {
-                fd.speculate_epoll(EPOLLIN);
-            }
-            return make_ready_future<size_t>(*r);
+        return do_with(::msghdr{}, iov, [&fd, this] (auto& mh, auto& iov) {
```
If you have a specialized io_completion subclass, it can host the msghdr so you can avoid the extra allocation.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1531,26 +1560,24 @@ class reactor_backend_uring final : public reactor_backend {
-            if (size_t(*r) == len) {
+            return queue_request(internal::io_request::make_send(fd.fd.get(), buffer, len, MSG_NOSIGNAL));
+        }).then([&fd, len] (size_t result) {
+            if (result == len) {
```
Maybe it's okay though.

tchaikov commented:
@avikivity hi Avi, thank you for your reviews and suggestions! i think i've addressed them all. could you take another look?

avikivity commented:
> @avikivity hi Avi, thank you for your reviews and suggestions! i think i've addressed them all. could you take another look?

I see you still call readable() directly.

tchaikov commented:
> > @avikivity hi Avi, thank you for your reviews and suggestions! i think i've addressed them all. could you take another look?
> 
> I see you still call readable() directly.

@avikivity because `read()` does not support `MSG_NOWAIT` flag. and `read_some(pollable_fd_state& fd, void* buffer, size_t len)` is only used by `seastar::fsnotifier`. assuming it's not on the critical path, it's not that bad. am i right?

avikivity commented:
> tested on a 8-core x86 machine with `httpd --smp 1` (release build)
and `wrk -c 128 -t 4`, the qps increased from 78750.08 (aio backend)
to 98570.76 (--reactor-backend io_uring).

Nice

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1136,6 +1136,10 @@ try_create_uring(unsigned queue_len, bool throw_on_error) {
+            IORING_OP_RECV,
+            IORING_OP_RECVMSG,
+            IORING_OP_SEND,
+            IORING_OP_SENDMSG,
```
But the kernel we run on can be older.

tchaikov replied:
ahh, right! i checked the linux source tree: 
- readv, writev were introduced in linux v5.1: https://github.com/torvalds/linux/blob/e93c9c99a629c61837d5a7fc2120cd2b6c70dbdd/include/uapi/linux/io_uring.h#L53-L55
- {send, recv}msg were introduced in linux v5.3: https://github.com/torvalds/linux/blob/4d856f72c10ecb060868ed10ff1b1453943fc6c8/include/uapi/linux/io_uring.h#L62-L63
- send and recv were introduced in linux v5.6: https://github.com/torvalds/linux/blob/7111951b8d4973bda27ff663f2cf18b663d15b48/include/uapi/linux/io_uring.h#L112-L113

these findings match the tags in https://unixism.net/loti/ref-liburing/submission.html .

will fallback to the aio implementation if a certain op does not exist. i understand the rationales but i think users/developers who want to use io_uring are more likely to use a newer kernel just to enjoy the new features. in my case, i'd even try out kernel 6.0 just for the zero copy send/recv. anyway, probably we can raise the bar in one years or two. as i think from the user's perspective, the difference between v5.1 and v5.3 or even v5.6 is not that large.

tchaikov replied:
just realized that read and write were added in v5.6. see https://unixism.net/loti/ref-liburing/submission.html#c.io_uring_prep_read . so we require linux 5.6 already . but i will add the machinery to detect the supported ops anyway. as i originally wanted to implement the [zero-copy transmission](https://lwn.net/Articles/879724/). in the long run, there are good chances that we support some bleeding edge ops, while user might fail to use Seastar io_uring backend just because his/her clients don't want to upgrade the kernel to linux v10.10!

tchaikov replied:
after a second thought, i think it'd be difficult to use a "downgraded" uring backend if we want to support the read and write ops. we would have to fallback to the aio backend for handling the ops in io_sink. so, can we make v5.6 the minimum linux kernel supported by the io_uring backend. the `reactor_backend_uring_v5{1,3,6}` change i just pushed is still useful, but only when we need to implement a feature only available in a newer kernel than v5.6. `IORING_OP_SENDMSG_ZC` support would be a good chance for using this mechinary. but probably not this time. i pushed it just in case. but personally, i don't really want to have this complexity at this moment.

avikivity replied:
I don't understand. Can't we make all the code added in this series optional? Then everything beyond the basic PREAD/PWRITE/POLL is detected and optional.

tchaikov replied:
the first batch got added to linux kernel were readv, writev, fsync, read_fixed, write_fixed, poll_add, and poll_remove. but our io_uring support already uses read, write and poll_add. in which, read and write were added in linux 5.6. also, we don't use readv or writev yet. so the minimum required linux version of our io_uring support is already linux 5.6.

avikivity replied:
I see. But we do use readv, for disk I/O.

avikivity replied:
Note, we could use READV instead of READ. But there's no need to be so strict and we can settle on 5.6 as our base. In any case io_uring's value becomes important with sendmsg/recvmsg.

tchaikov replied:
ahh, right. we do use readv. i missed it.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1313,11 +1317,16 @@ class reactor_backend_uring final : public reactor_backend {
+            return true;
         }
+        return false;
+    }
```
Let's make this a member of pollable_fd_state for cleanliness.

The name should also indicate that it clears the speculation. Perhaps pollable_fd_state::get_and_clear_speculation(int events)?

tchaikov replied:
how about `take_speculation()`?

avikivity replied:
ok, as long as it's explained.

tchaikov replied:
added comment to explain the API and the rationales.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1478,51 +1495,143 @@ class reactor_backend_uring final : public reactor_backend {
+            }
+        };
+        auto desc = std::make_unique<read_completion>(fd, iov);
+        auto req = internal::io_request::make_recvmsg(fd.fd.get(), desc->msghdr(), MSG_DONTWAIT);
```
MSG_DONTWAIT is wrong here, we do want to wait, otherwise we can get EAGAIN.

Not sure how this worked for you, since http should never speculate as readable. Maybe I don't understand how MSG_DONTWAIT interacts with io_uring correctly?

tchaikov replied:
ahh. right! the uring call should be *blocking*. it was a silly replace-them-all bug. let me retest.

tchaikov replied:
> Not sure how this worked for you, since http should never speculate as readable. Maybe I don't understand how MSG_DONTWAIT interacts with io_uring correctly?

i came across this topic the other day. my understanding is that io_uring does return EAGAIN if the fd is opened with `O_NONBLOCK`. and `MSG_DONTWAIT` works like we expect. see https://patchwork.kernel.org/project/linux-fsdevel/patch/20190629161204.27226-2-axboe@kernel.dk/ .

guess i was just lucky when testing my change, as both the server (httpd) and the client (wrk) are located on the same machine.

avikivity replied:
Yes, we should open the file without SOCK_NONBLOCK if uring sendmsg is available (and carefully make sure we always use SOCK_NONBLOCK everywhere).

tchaikov replied:
@avikivity i was wrong! i realized that io_uring *never* returns -EAGAIN when reading the commit message of https://github.com/torvalds/linux/commit/d7718a9d25a61442da8ee8aeeff6a0097f0ccfd6 .

so we don't need to care about the `SOCK_NONBLOCK` anymore. io_uring takes care of this already. and it is even more efficient after linux v5.7 !

tchaikov replied:
this also explains why `internal::io_request::make_recvmsg(fd.fd.get(), desc->msghdr(), MSG_DONTWAIT)` worked last week end.

avikivity replied:
That's very convenient for us. Please ask it to be documented on the uring mailing list, this is not intuitive and in order to trust it not to change it's better if it's documented.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1531,31 +1640,89 @@ class reactor_backend_uring final : public reactor_backend {
+            auto r = fd.fd.sendmsg(&mh, MSG_NOSIGNAL | MSG_DONTWAIT);
+            if (r) {
+                if (size_t(*r) == p.len()) {
+                    fd.speculate_epoll(EPOLLOUT);
```
This can throw. Please check that the callers catch the exception and return a future with an exception (for the other calls too).

avikivity replied:
By "this" I meant the call to sendmsg().

tchaikov replied:
yeah, `fd.fd.sendmsg()` throws. but if we check the diff in this commit (https://github.com/scylladb/seastar/pull/1235/commits/1a9baa61eeb1358f2183eef2407ad5032e313aaf#diff-33e3c9bf9758fbe5445e34b395c6f07d6b1ea890ba224e605a7e7860f3b5064aR1646) , i made no change in this area. the existing callers already (or should already) catch the expected exceptions. 

> return a future with an exception (for the other calls too).

IIUC, if a continuation body throws, we already set the state in the corresponding promise with the exception. right? in other words, since the underlying future/promise framework has taken care of this, we don't need to do extra work, right?

tchaikov replied:
thanks. i see now. will do something like `futurize_invoke()`.

tchaikov replied:
hmm, thought i create a clever wrapper do implement this. but the code flow with early returns makes it difficult. so i am just putting it in the unrolled form. but the upside is that it is easier to read. =)

avikivity replied:
I think it's fine to be unrolled here.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1486,10 +1492,98 @@ class reactor_backend_uring final : public reactor_backend {
+            void complete(size_t fd) noexcept final {
+                _listenfd.speculate_epoll(EPOLLIN);
+                pollable_fd pfd(file_desc::from_fd(fd), pollable_fd::speculation(EPOLLOUT));
+                _result.set_value(std::move(pfd), std::move(_sa));
```
Do we have tests that server_socket::abort_accept() causes server_socket::accept() to return? If not please add a test.

The code looks like it should work, but better to be sure.

tchaikov replied:
yeah. actually, i added this just to fix the unexpected invalid_argument (errno 22) exception thrown when i stopped the httpd server.

now since the accept+connect patch will be dropped in this PR. i will continue working on it in another one.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1486,10 +1492,98 @@ class reactor_backend_uring final : public reactor_backend {
+            }
+        };
+        auto desc = std::make_unique<accept_completion>(listenfd);
+        return readable_or_writeable(listenfd).then([&listenfd, desc = std::move(desc), this]() mutable {
```
Without speculating here, this will be slower, not faster, since speculation will usually be false, we'll go into the kernel to poll, then go to the kernel again for the accept.

I suggest to drop this patch and come back to it later. Even if it gives a huge speedup it will not be noticeable since only a tiny fraction of the cpu is spent on accepts. Let's concentrate on send/recv first.

tchaikov replied:
sure. will create another PR after the send/recv one lands.

avikivity commented:
Looks very good. I suggest to park connect/accept and get send/recv in first.

After this gets in, please send a patch to make uring the preferred backend.

tchaikov commented:
changelog

- added a chain of inheritance of io_uring backends
- added runtime detection of supported ops
- use a factory method to choose different io_uring backend based on the detection result
- dropped the wrong `MSG_DONTWAIT` flags passed to io_uring ops
- dropped the accept+connect patch
- move `is_speculated()` into `pollable_fd_state` class. and rename it to `take_speculation()`.

tchaikov commented on src/core/reactor_backend.cc:
```diff
@@ -1211,7 +1220,10 @@ detect_io_uring() {
 }
 
-class reactor_backend_uring final : public reactor_backend {
+class reactor_backend_uring_v51 : public reactor_backend {
```
@avikivity hi Avi, could you take a look at https://github.com/scylladb/seastar/pull/1235#discussion_r989727504 ?

tchaikov commented:
changelog

- revert the `reactor_backend_uring_v5{1,3,6}` change
- futurize the exception thrown by function calls like `fd.fd.send()`
- reorder `required_ops` in `try_create_uring` and add comments on when these ops are added.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1134,12 +1134,16 @@ try_create_uring(unsigned queue_len, bool throw_on_error) {
+            IORING_OP_READ,     // linux 5.6
+            IORING_OP_WRITE,
+            IORING_OP_SEND,
+            IORING_OP_RECV,
```
Why are you making these required?

tchaikov replied:
because read and write are required. and both send and recv were added in the exact same linux version (5.6) when read and write were added. so we get send and recv for free..

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1478,51 +1493,149 @@ class reactor_backend_uring final : public reactor_backend {
-            if (!r) {
-                return read_some(fd, iov);
+            try {
+                auto r = fd.fd.recvmsg(&mh, MSG_DONTWAIT);
```
I think it's confusing that some read_some() overloads use read(), and some read_some() overloads use recvmsg(). Changing an overload can cause a crash (for the inotify use case, if we switch to sendmsg) or reduced performance (if we switch to read).

I think we need separate calls in reactor and reactor_backend for read/write/sendmsg/recvmsg, and then to switch socket calls to read/write.

Ah, but I see the problem already exists in master, so I won't require it as a prerequisite.

tchaikov replied:
it is. i had a hard time getting my head around this. will fix this in a follow up change.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1478,51 +1493,149 @@ class reactor_backend_uring final : public reactor_backend {
+                    return make_ready_future<temporary_buffer<char>>(std::move(buffer));
+                }
+            } catch (const std::system_error& e) {
+                return make_exception_future<temporary_buffer<char>>(e);
```
You can use `catch (...)` and `current_exception_as_future()`, they're safer. In particular copying an exception caught by reference can lead to object slicing.

tchaikov replied:
okay. wanted to make it explicit by capturing only the expected exception type. will change in the next revision.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1478,51 +1493,149 @@ class reactor_backend_uring final : public reactor_backend {
+            } catch (const std::system_error& e) {
+                return make_exception_future<temporary_buffer<char>>(e);
             }
-            buffer.trim(*r);
```
Did we lose the trim() call?

tchaikov replied:
no. it is moved into `read_completion::complete(size_t bytes)`, see https://github.com/scylladb/seastar/pull/1235/commits/aff2fa69c449a7a44f8e8545f51b4581afbb9c08#diff-33e3c9bf9758fbe5445e34b395c6f07d6b1ea890ba224e605a7e7860f3b5064aR1614

avikivity replied:
But this read_completion::complete() isn't called in this path.

tchaikov replied:
ahh, my bad. fixed in the latest revision.

avikivity commented on src/core/reactor_backend.hh:
```diff
@@ -212,6 +212,10 @@ public:
     virtual pollable_fd_state_ptr make_pollable_fd_state(file_desc fd, pollable_fd::speculation speculate) = 0;
+
+    virtual bool do_blocking_io() const {
+        return false;
```
This should return true or false depending on the detected capabilities.

avikivity replied:
Since it's okay to settle on a 5.6 baseline, it can be unconditionally false.

tchaikov replied:
ack.

tchaikov commented:
changelog

- add the missing `buffer.trim(*r)` in `read_some(pollable_fd_state& fd, internal::buffer_allocator* ba)` 's optimistic code path
- use `catch (...)` and `current_exception_as_future()` to futurize the exception thrown by `file_desc` 's POSIX call methods
- add doxygen document to `pollable_fd_state::speculate_epoll()` and `pollable_fd_state::take_speculation()`.

tchaikov commented:
```console
# apps/httpd/httpd --smp 1
$ ./wrk -c 128 -t 4 http://localhost:10000/
Running 10s test @ http://localhost:10000/
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.52ms  151.61us   4.64ms   82.54%
    Req/Sec    21.05k     1.09k   24.49k    86.25%
  837710 requests in 10.00s, 109.45MB read
Requests/sec:  83759.32
Transfer/sec:     10.94MB
```

```console
# apps/httpd/httpd  --reactor-backend io_uring --smp 1
$ ./wrk -c 128 -t 4 http://localhost:10000/
Running 10s test @ http://localhost:10000/
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.48ms  575.17us  44.35ms   99.39%
    Req/Sec    21.93k     2.37k   25.59k    67.00%
  873133 requests in 10.00s, 114.08MB read
Requests/sec:  87302.20
Transfer/sec:     11.41MB
```

tchaikov commented:
> Looks very good. I suggest to park connect/accept and get send/recv in first.
> 
> After this gets in, please send a patch to make uring the preferred backend.

sure. will do.

tchaikov commented:
@avikivity in case you missed it. see https://github.com/scylladb/seastar/pull/1235#discussion_r991829176

avikivity commented:
> ```
> # apps/httpd/httpd --smp 1
> $ ./wrk -c 128 -t 4 http://localhost:10000/
> Running 10s test @ http://localhost:10000/
>   4 threads and 128 connections
>   Thread Stats   Avg      Stdev     Max   +/- Stdev
>     Latency     1.52ms  151.61us   4.64ms   82.54%
>     Req/Sec    21.05k     1.09k   24.49k    86.25%
>   837710 requests in 10.00s, 109.45MB read
> Requests/sec:  83759.32
> Transfer/sec:     10.94MB
> ```
> 
> ```
> # apps/httpd/httpd  --reactor-backend io_uring --smp 1
> $ ./wrk -c 128 -t 4 http://localhost:10000/
> Running 10s test @ http://localhost:10000/
>   4 threads and 128 connections
>   Thread Stats   Avg      Stdev     Max   +/- Stdev
>     Latency     1.48ms  575.17us  44.35ms   99.39%
>     Req/Sec    21.93k     2.37k   25.59k    67.00%
>   873133 requests in 10.00s, 114.08MB read
> Requests/sec:  87302.20
> Transfer/sec:     11.41MB
> ```

Performance improvement is much lower. Do we know why?

It could be that the speculation fixes caused all the writes to go via sendmsg() and not uring. This improves latency but reduces throughput.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1478,51 +1493,150 @@ class reactor_backend_uring final : public reactor_backend {
+            };
+            auto desc = std::make_unique<read_completion>(fd, len);
+            auto req = internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false);
+            return submit_request(std::move(desc), std::move(req));
```
I think we need to keep this as the original loop, since the new code takes two round-trips in the ring, so has higher latency.

Although, since it's just for inotify, it doesn't matter much.

tchaikov replied:
as io_uring `read` op will retry automatically, i think we can dispense with the `readable(fd).then` part. so in the worst case, just one round trip would suffice.

anyway, will add this change in a follow-up PR.

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -1531,31 +1645,93 @@ class reactor_backend_uring final : public reactor_backend {
-        });
+            void complete(size_t bytes) noexcept final {
+                if (bytes == _to_write) {
+                    _fd.speculate_epoll(EPOLLIN);
```
This should be EPOLLOUT, no?

tchaikov replied:
ahhh, true. fixed.

avikivity commented:
Please expand the cover letter (esp. with test results)

tchaikov commented:
changelog

- added the wrk+httpd test results in the commit message and in cover letter.
- dropped the `read_some(pollable_fd_state& fd, void* buffer, size_t len)` io_uring implementation.
- s/EPOLLIN/EPOLLOUT/ in `write_some(pollable_fd_state& fd, net::packet& p)` io_uring implementation.

tchaikov commented:
> 
> Performance improvement is much lower. Do we know why?
> 
> 
> It could be that the speculation fixes caused all the writes to go via sendmsg() and not uring. This improves latency but reduces throughput.

i don't have a theory explaining this. actually the wrk test result is not very consistent. i just picked the one with lowest stddev. but in general the io_uring's throughput is always better than that of aio backend. and the average latency is a little bit better than aio, despite that its stddev is always 2 times to 4 times of that of aio. yeah, guess it's because that some times the I/Os are handled by POSIX call and some times they are processed by io_uring.

i think we need to hammer the server harder with two machines to get a more realistic benchmark.

as a note to myself:

- [ ] to document the automatic retry behavior in liburing
- [ ] to add recv+connect to io_uring backend
- [ ] to add `read_some(pollable_fd_state& fd, void* buffer, size_t len)` back to io_uring backend
- [ ] to make uring the preferred backend.
- [ ] to use separate calls in reactor and reactor_backend for read/write/sendmsg/recvmsg, and then to switch socket calls to read/write.
